content/200-concepts/100-components/02-prisma-client/052-working-with-generated-types/index.mdx

```diff
@@ -0,0 +1,101 @@
+---
+title: 'Advanced usage of generated types'
+metaTitle: 'Advanced usage of generated TypeScript types (Reference)'
+metaDescription: 'Prisma Client provides full type safety for queries, even for partial queries or included relations. This page explains how to leverage the generated types.'
+---
+
+<TopBlock>
+
+The generated code for Prisma Client contains several helpful types that you can use to make your application more type-safe. This page describes patterns for leveraging some of the generated types.
+
+By importing the `Prisma` namespace from `@prisma/client` you can access types related to your defined models.
+
+The following example composes a `select` param outside of the queries context using the [`validator`](#prisma-validator) function, it passes in a generated type (`UserSelect`) meaning only fields available on a `select` can be used.
+
+```ts
+// Import the `Prisma` namespace from the `@prisma/client` package
+import { PrismaClient, Prisma } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+// Pass in the UserSelect type assertion
+const selectId = Prisma.validator<Prisma.UserSelect>()({
+  id: true,
+})
+
+const data = await prisma.user.findFirst({ select: selectId })
+```
+
+> **Note**: You can find all type definitions inside the `index.d.ts` file inside the [generated `.prisma/client` folder](working-with-prismaclient/generating-prisma-client#the-prismaclient-npm-package).
+
+</TopBlock>
+
+### <inlinecode>Prisma.validator</inlinecode>
+
+The `validator` is a function that exists on the Prisma namespace. It accepts a generated type and returns an object that fits that types shape. It helps you create re-usable query parameters based on your schema models while making sure that the objects you create are valid.
```

Comment:
> It accepts a generated type and returns an object that fits that types shape

Looking at the signature again, I think this is not quite accurate:

```ts
export function validator<V>(): <S>(select: Exact<S, V>) => S;
```

based on the signature, I _think_ the following statements would be true:

- it accepts an object `select` of type `Exact<S, V>` (here it would be helpful to have a simple description of what the `Exact<S, V>` is and what it does)
- it takes two generic type parameter `V` and `S`
- it uses `S` for a type assertion on `select`
- it returns an object of type `S`

I think we should spend some time here explaining what is going on. At least on a high-level, all of this should also be understandable by a TS _beginner_ (in the sense that they at least understand what this function is used for although they might not understand all the details). Plus I think we should link out to the TS docs when referencing concepts like [Type Assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) and [Generics](https://www.typescriptlang.org/docs/handbook/2/generics.html).